design/paging.md

```diff
@@ -0,0 +1,119 @@
+# Paging
+
+## Document purpose
+
+Describe options for paging through collections data (although this discussion is generally
+applicable to any data stored in a database that is too large to return to the user at once)
+and their pros and cons.
+
+Paging is often an afterthought or considered a non-issue, but this is not the case, and there
+is no ideal solution that the author is aware of. Database schemas and APIs may need modification
+to page effectively and efficiently if not designed with that in mind.
+
+## Nomenclature
+
+* `documents` can be read as documents in in a document based database, like MongoDB or ArangoDB,
+  or rows in a RDBMS, like MySQL or Postgres.
+
+## Options
+
+### Skip and limit / offset and limit
+
+This is the standard paging solution of providing a count of documents to skip and a count
+of documents to return.
+
+#### Pros
+
+* Can page to any location in the data
+* Easily understood by users
+
+#### Cons
+
+* `O(n^2)` time complexity for most database implementations
+  * This means that in larger data sets deep paging gets progressively, exponentially slower
+  * A unlimited skip and limit implementation in the KBase Workspace caused MongoDB to slow to
+    a literal halt due to UI paging code written by someone who should've known better
+    * Skip / limit was removed after that
+  * Often implemented with a maximum skip allowed to prevent slow paging
+    * Elasticsearch's limit is 10K by default
+    * This means that only the data at the front of the sort can be paged
+
+### `start_after` with limit
+
+Provide a string or numerical value after which paging should start. Get the next start after
+value from the last value of the current page.
+
+E.g. if the current page runs from `Ababdeh` to `abaciscus` the `start_after` value for the next
+page would be `abasciscus`.
+
+#### Pros
+
+* Easily understood by users if implemented as a simple "next page" button
+* Users can jump to a particular value rather than having to search through pages if they
+  are allowed to specify the `start_after` value
+  * Essentially a filter used for paging
+* Prior `start_after` values can be cached by the front end to allow for returning to a
+  previous page
+* At worst linear time complexity
+
+#### Cons
+
+* Can only navigate to the next page after the current page. Navigating to page 10 from page
+  1 would require 9 full page loads.
+* May not work well on non-unique fields
+  * If a page has more than a page size's number of identical values in the field being paged over,
```

Comment:
Couldn't this be solved by having an numbered index per sortable column? That might be a bit bloated though...

Reply:
That's essentially solution #3

Reply:
If you're going to do that you might was well use that numbered index as the `start_after` value

Reply:
Added to the doc